src/Polysemy/Final/Writer.hs

```diff
@@ -0,0 +1,114 @@
+{-# LANGUAGE BangPatterns, TupleSections #-}
+module Polysemy.Final.Writer where
+
+import Control.Concurrent.STM
+
+import Polysemy
+import Polysemy.Final
+import Polysemy.Writer
+import Control.Exception
+
+--------------------------------------------------------------------
+-- | Transform a 'Writer' effect to atomic operations
+-- over a 'TVar'.
+runWriterTVar :: (Member (Final IO) r, Monoid o)
+              => TVar o
+              -> Sem (Writer o ': r) a
+              -> Sem r a
+runWriterTVar tvar = runWriterTVarAction $ \o -> do
+  s <- readTVar tvar
+  writeTVar tvar $! s <> o
+{-# INLINE runWriterTVar #-}
+
+-- TODO(KingoftheHomeless): Make this mess more palatable
+--
+-- 'interpretFinal' is too weak for our purposes, so we
+-- use 'interpretH' + 'withWeaving'.
+runWriterTVarAction :: forall o r a
+                          . (Member (Final IO) r, Monoid o)
```

Comment:
This could be generalized to `Member (Final m) r, MonadIO m, MonadMask m`, but I doubt that's worth it.

Reply:
Let's keep the mtl stuff as far away as possible :)

Reply:
Sure.